### PR TITLE
When refreshing a stylesheet, schedule a report CSS values once the sheet is loaded

### PR DESCRIPTION
**Pull request by cpsievert**
Should fix timing issues with https://connect.rstudioservices.com/explore_your_weather

### CODE REVIEW

File: srcts/src/main.ts
Comment (cpsievert) [inline srcts/src/main.ts:4071]: Instead of `Shiny.bindAll` lets just send the CSS info (`sendImageSize`) and throttle calls to it.

**Comment by wch** [comment]
For future reference, with this PR, when I load a Shiny app, it calls `sendImageSize()` 403 times.

image


Changing the app from light to dark mode results in about 400 more calls to that function.

**Comment by wch** [comment]
It looks like the debouncing reduces the number of calls to `doSendImageSize()` to 101 times when the app is loaded. But that still seems pretty expensive, considering that each call to `doSendImageSize()` does a fair bit of work, and triggers `shiny:visualchange`:

https://github.com/rstudio/shiny/blob/16e0d9e355713e6765d69c9203cb29ee2af4eeb7/srcts/src/main.ts#L6843-L6878

**Comment by wch** [comment]
@cpsievert I recall that there can be a delay between when the CSS is loaded and when it is actually applied. I also remember that the delay can vary. Do you remember why it can take so long -- up to 10 seconds?

I'm considering the following:

* Use a MutationObserver to re-send image size information when an image/plot output changes size (instead of listening for the `window` `resize` event). This would also trigger the `shiny:visualchange` event. We'd need some sort of workaround for IE11, but it wouldn't need to be perfect.
* To detect CSS style changes is trickier, since there's no observer that detects those changes. I wonder if, in the `onload` call back for CSS file, we can set some arbitrary CSS style property via JS, and then trigger a callback after that is applied. If the CSS style set by JS always takes effect after the file-based CSS styles, that could work. But we'll have to test to make sure.

**Comment by cpsievert** [comment]
> Do you remember why it can take so long -- up to 10 seconds?

10 seconds is a somewhat arbitrary decision. Same with the 0.1sec interval (which combined is leading to the 101 calls to `doSendImageSize()`). Although, it's important to keep the interval somewhat small so that you're not waiting forever to get an update. Maybe it's make sense to implement some sort of exponential back-off and/or clear the interval if the value has changed (I'd probably need help implementing the later)

**Comment by wch** [comment]
With the new method I've implemented, `doSendImageSize` gets called 3 times on load, and 2 times when switching styles.

image

File: srcts/src/main.ts
Comment (cpsievert) [inline srcts/src/main.ts:4057]: Would be good to have a comment here about this working because (as far as we know), in this case, styles are applied syncronously

File: srcts/src/main.ts
Comment (cpsievert) [inline srcts/src/main.ts:4123]: If we wanted to be extra careful, maybe add the `!important` here as well?

File: srcts/src/main.ts
Comment (cpsievert) [inline srcts/src/main.ts:4136]: I think we should be using `removeSheet()` here -- this won't necessarily disable the stylesheet

**Comment by cpsievert** [comment]
LGTM once merge conflicts are addressed, thanks!

**Review by wch — APPROVED** [review]

**Pull request merged**
